### PR TITLE
Add getUnregisteredEnrollments function in EnrollmentManager

### PR DESCRIPTION
**Pull request by TrustHenry**
Get the unregistered enrollments in the block
And this is arranged in ascending order with the utxo_key.

For #493

### CODE REVIEW

File: source/agora/consensus/EnrollmentManager.d [hunk 1]
```diff
@@ -423,6 +423,35 @@ public class EnrollmentManager
...
+
+        Params:
+            enrolls = will contain the unregistered enrollments data if found
+
+        Returns:
+            Return true if the unregistered enrollment data was found
+
+    ***************************************************************************/
+
+    public bool getEnrollments (out Enrollment[] enrolls) @trusted
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: When you use `out`, memory is not reused. Because `out` will initialize the array to its default value, which is `null`.
In order to reuse memory, you should use `ref`.

File: source/agora/consensus/EnrollmentManager.d [hunk 2]
```diff
@@ -423,6 +423,35 @@ public class EnrollmentManager
...
+    {
+        enrolls.length = 0;
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d:436]: If your goal here is to reuse memory, you should use `assumeSafeAppend`

File: source/agora/consensus/EnrollmentManager.d [hunk 3]
```diff
@@ -423,6 +423,35 @@ public class EnrollmentManager
         return false;
     }
 
+    /***************************************************************************
+
+        Get the unregistered enrollments in the block
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d:422]: The name should reflect that it's the unregistered enrollments, not just enrollments.

File: source/agora/consensus/EnrollmentManager.d [hunk 4]
```diff
@@ -423,6 +423,35 @@ public class EnrollmentManager
...
+        enrolls.length = 0;
+        auto results = this.db.execute("SELECT val FROM validator_set" ~
+            " WHERE enrolled_height is null");
+
+        foreach (row; results)
+            enrolls ~= deserializeFull!Enrollment(row.peek!(ubyte[])(0));
+
+        if (enrolls.length > 0)
+        {
+            enrolls.sort!("a.utxo_key < b.utxo_key");
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: I think we could move this to the query, couldn't we ?

File: source/agora/consensus/EnrollmentManager.d [hunk 5]
```diff
@@ -525,6 +554,11 @@ unittest
     assert(man.addEnrollment(enroll2));
     assert(man.getEnrollmentLength() == 2);
 
+    Enrollment[] enrolls;
+    man.getEnrollments(enrolls);
+    assert(enrolls.length == 2);
+    assert(enrolls[0].utxo_key < enrolls[1].utxo_key);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: I think a full test would include:
- 3 enrollments, added in (utxo key) order;
- 3 enrollments, added in reverse order;

At the moment the sortedness is not really tested. It should also be tested on the block validation level, but maybe you want to keep this for another PR ?

File: source/agora/consensus/EnrollmentManager.d (hunk 4 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: I thought I could solve it with Query.
In DB, the key is saved as Byte[]
Then it should encode again and order it into Query.
Would it be better to do this?

Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: So the issue is that SQLite is not able to ORDER BY on binary data ?

Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: If it's the same sort order of binary data, it's easy to fix with order by.

File: source/agora/consensus/EnrollmentManager.d (hunk 5 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Yes, I'll add a test at the block level in another PR to be added.

File: source/agora/consensus/EnrollmentManager.d (hunk 3 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d:422]: I changed the function name `getUnregisteredEnrollments`

File: source/agora/consensus/EnrollmentManager.d (hunk 2 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d:436]: I add `assumeSafeAppend(enrolls)`

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: applied the suggest

File: source/agora/consensus/EnrollmentManager.d (hunk 4 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Binary data sorted by ORDER BY key ASC by SQLite

```
[14, 164, 193, 71, 177, 203, 141, 122, 96, 154, 232, 88, 253, 37, 3, 70, 159, 11, 43, 84, 191, 254, 138, 190, 72, 254, 70, 251, 59, 120, 47, 51, 45, 21, 81, 18, 208, 26, 28, 151, 112, 225, 205, 2, 10, 168, 89, 99, 223, 16, 73, 44, 166, 34, 182, 144, 194, 52, 250, 97, 16, 95, 104, 163]
[123, 195, 91, 196, 99, 236, 96, 3, 9, 136, 102, 72, 138, 254, 100, 252, 104, 23, 220, 171, 226, 215, 168, 54, 50, 240, 204, 203, 161, 107, 63, 198, 195, 254, 195, 185, 226, 63, 38, 197, 150, 40, 255, 117, 25, 242, 86, 84, 62, 31, 35, 78, 3, 232, 114, 79, 30, 182, 134, 111, 125, 171, 231, 81]
[232, 104, 205, 80, 68, 5, 170, 214, 76, 66, 173, 149, 48, 25, 96, 205, 59, 51, 251, 106, 84, 10, 53, 69, 128, 54, 153, 28, 78, 58, 218, 53, 19, 202, 39, 184, 66, 75, 77, 115, 153, 133, 128, 174, 41, 199, 175, 192, 161, 104, 94, 240, 13, 66, 5, 37, 21, 131, 133, 56, 137, 193, 218, 237]
```
deserialize from Binary data 

```
0xa3685f1061fa34c290b622a62c4910df6359a80a02cde170971c1ad01251152d332f783bfb46fe48be8afebf542b0b9f460325fd58e89a607a8dcbb147c1a40e
0x51e7ab7d6f86b61e4f72e8034e231f3e5456f21975ff2896c5263fe2b9c3fec3c63f6ba1cbccf03236a8d7e2abdc1768fc64fe8a486688090360ec63c45bc37b
0xeddac189388583152505420df05e68a1c0afc729ae808599734d4b42b827ca1335da3a4e1c99368045350a546afb333bcd60193095ad424cd6aa054450cd68e8
```

Not sorted in order of desired UTXOkey hash.

Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Oh, I see. The binary data for SQLlite was little endian.

Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: I made a new function to reverse bytes before putting it in to DB.
It is also used in SELECT.
@Geod24 Please check to see if it's okay.

File: source/agora/consensus/EnrollmentManager.d [hunk 6]
```diff
@@ -451,6 +483,26 @@ public class EnrollmentManager
...
+
+        Return:
+            Reverse bytes and return to the new ubyte[]
+
+    ***************************************************************************/
+
+    private static ubyte[] reverseUbytes (ref const Hash data) @trusted nothrow
+    {
+        ubyte[] bytes;
+        foreach (val; data[].dup.reverse)
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Why the `dup` here ? It allocates data for no reason. If `reverse` is in place, then you can just `return data[].dup.reverse`, no need for the loop. If it does not allocate, then why the `dup` ?

File: source/agora/consensus/EnrollmentManager.d [hunk 7]
```diff
@@ -389,7 +390,7 @@ public class EnrollmentManager
     public bool hasEnrollment (const ref Hash enroll_hash) @trusted
     {
         auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM validator_set " ~
-            "WHERE key = ?)", enroll_hash[]);
+            "WHERE key = ?)", reverseUbytes(enroll_hash));
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Ultimately, the library should accept either ranges, or recognize `toString`. Could you check what code is taken here ?

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/consensus/EnrollmentManager.d (hunk 6 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Oh, that's right. applied the suggestion.

File: source/agora/consensus/EnrollmentManager.d [hunk 8]
```diff
@@ -451,6 +483,23 @@ public class EnrollmentManager
...
+            data = hash of target to reverse
+
+        Return:
+            Reverse bytes and return to the new ubyte[]
+
+    ***************************************************************************/
+
+    private static ubyte[] reverseUbytes (ref const Hash data) @trusted nothrow
+    {
+        return data[].dup.reverse;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: What is this function for? The commit message doesn't explain it. Why is the byte array reversed?

Also, this should use a buffer instead. Like a `static ubyte[]`, set its length to `data[]` and then copy the data over with `buffer[] = data[]`.

File: source/agora/consensus/EnrollmentManager.d [hunk 9]
```diff
@@ -423,6 +424,37 @@ public class EnrollmentManager
...
+
+        enrolls.length = 0;
+        assumeSafeAppend(enrolls);
+        auto results = this.db.execute("SELECT val FROM validator_set" ~
+            " WHERE enrolled_height is null ORDER BY hex(key) ASC");
+
+        foreach (row; results)
+            enrolls ~= deserializeFull!Enrollment(row.peek!(ubyte[])(0));
+
+        if (enrolls.length > 0)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: Just `return enrolls.length > 0` will do. :)

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d]: Actually..

@Geod24 I feel like a function should not have a success/failure return code if you can already determine at the call site if the `enrolls` array is empty or not. What do you think?

Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: Yeah it should probably just `return ref`.

File: source/agora/consensus/EnrollmentManager.d (hunk 7 above)
Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: Ping @TrustHenry

Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: This is a library for use only here.
If we publicize this, better add a function in Hash.
Am I right?

File: source/agora/consensus/EnrollmentManager.d (hunk 8 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: > What is this function for? The commit message doesn't explain it. Why is the byte array reversed?

Added commit message description.

> Also, this should use a buffer instead. Like a static ubyte[], set its length to data[] and then copy the data over with buffer[] = data[].

I added a buffer.

File: source/agora/consensus/EnrollmentManager.d (hunk 9 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: I changed the return type.

File: source/agora/consensus/EnrollmentManager.d (hunk 7 above)
Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: By "the library" I mean d2sqlite.

Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Oh, I will check it.

**Comment by TrustHenry** [comment]
I changed the key column to TEXT
I changed it completely by using enroll.utxo_key.toString()
Ready for review.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by TrustHenry** [comment]
@Geod24  ping.

File: source/agora/consensus/EnrollmentManager.d [hunk 10]
```diff
@@ -451,6 +479,7 @@ public class EnrollmentManager
         catch (Exception ex)
         {}
     }
+
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Unrelated

File: source/agora/consensus/EnrollmentManager.d [hunk 11]
```diff
@@ -525,14 +554,26 @@ unittest
...
+    auto utxo_hash3 = utxo_hashes[2];
+    Enrollment enroll3;
+    assert(man.createEnrollment(utxo_hash3, enroll3));
+    assert(man.addEnrollment(enroll3));
+    assert(man.getEnrollmentLength() == 3);
+
+    Enrollment[] enrolls;
+    man.getUnregisteredEnrollments(enrolls);
+    assert(enrolls.length == 3);
+    assert(enrolls[0].utxo_key < enrolls[1].utxo_key);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: You can just use https://dlang.org/phobos/std_algorithm_sorting.html#isStrictlyMonotonic

File: source/agora/consensus/EnrollmentManager.d [hunk 12]
```diff
@@ -543,4 +584,22 @@ unittest
...
+    Enrollment[] ordered_enrollments;
+    ordered_enrollments ~= enroll;
+    ordered_enrollments ~= enroll2;
+    ordered_enrollments ~= enroll3;
+    ordered_enrollments.sort!("a.utxo_key < b.utxo_key");
+    foreach (ordered_enroll; ordered_enrollments)
+        assert(man.addEnrollment(ordered_enroll));
+    man.getUnregisteredEnrollments(enrolls);
+    foreach (idx, get_enroll; enrolls)
+        assert(ordered_enrollments[idx] == get_enroll);
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Likewise, just call https://dlang.org/phobos/std_algorithm_sorting.html#isStrictlyMonotonic

**Review by Geod24 — COMMENTED** [review]
Made a couple comments, otherwise LGTM

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Addressed the review directly

**Comment by Geod24** [comment]
And rebased on master since Drey's recent PR conflicted with this.

File: source/agora/consensus/EnrollmentManager.d (hunk 11 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: I knew this.
It was used in #480PR.
I will improve and write a new PR.

File: source/agora/consensus/EnrollmentManager.d (hunk 12 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: ^

File: source/agora/consensus/EnrollmentManager.d (hunk 10 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: I will improve and write a new PR.

File: source/agora/consensus/EnrollmentManager.d (hunk 11 above)
Comment (TrustHenry) [reply source/agora/consensus/EnrollmentManager.d]: Oh, Thanks.
